### PR TITLE
bip-340: recreate batch verify speedup graph w/ latest libsecp256k1

### PR DESCRIPTION
**Pull request by jonasnick**
Since creating that graph there have been significant improvements in
libsecp256k1 to the speed of certain operations, such as making use of the
endomorphism by default and the safegcd algorithm. This new graph shows the
batch verification speedup in light of these improvements.

There are also some minor improvements to the graph itself. In particular,
the (1,1) datapoint was removed because it is not part of the estimation for the
fitted line.

This commit also contains the Makefile, shell script and gnuplot file that were
used to create the graph.

CC @sipa @real-or-random

### CODE REVIEW

**Comment by sipa** [comment]
Trying to run your benchmark code I get:

```sh
$ make
cat raw_output.txt | grep -v "schnorrsig_batch_verify_1" | awk 'match($0,                         /schnorrsig_batch_verify_(.*):.*avg (.*)us /, a) {print a[1] " " a[2]}' >                         batch.dat                                                                         awk: line 1: syntax error at or near ,
```

**Comment by jonasnick** [comment]
@sipa Hm, I can't reproduce this. `awk --version` is `GNU Awk 5.1.0` on my end.

**Comment by sipa** [comment]
@jonasnick Ok, apparently I was using "mawk" instead of "gawk".

I redid the benchmarks myself (Intel i7-7820HQ, clock fixed at 2.6 Ghz) with higher granularity and higher iteration count, and got this graph:

speedup-batch

I'm not sure a linear interpolation is right...

File: bip-0340/speedup-batch/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+bench_output.txt: bench.sh
+	SECP256K1_BENCH_ITERS=100000 ./bench.sh bench_output.txt
+
+batch.dat: bench_output.txt
+	cat bench_output.txt | grep -v "schnorrsig_batch_verify_1" | awk 'match($$0, /schnorrsig_batch_verify_(.*):.*avg (.*)us /, a) {print a[1] " " a[2]}' > batch.dat
```
Comment (sipa) [inline bip-0340/speedup-batch/Makefile:5]: This is ignoring all numbers where the batch size starts with decimal digit "1". By using `grep -v "schnorrsig_batch_verify_1:"` that is avoided.

Comment (jonasnick) [reply bip-0340/speedup-batch/Makefile]: good catch, fixed

**Comment by jonasnick** [comment]
I'll rerun the experiment with similar granularity. That many data points makes the fitted line useless anyway.

**Comment by sipa** [comment]
New graph with much higher scratch space:

speedup-batch

**Review by 0racl3z — APPROVED** [review]

**Comment by jonasnick** [comment]
Before creating a new graph, we should take into account 1) https://github.com/sipa/bips/issues/219 and 2) new estimated optimal pippenger threshold/windows.

**Comment by sipa** [comment]
@jonasnick Also, the optimal pippenger thresholds/windows may be affected by the use of 128-bit randomizers?

**Comment by 0racl3z** [comment]
Im supposed to be using tails aren't i..😔

**Comment by luke-jr** [comment]
What is the status here? It's marked as a Draft...

**Comment by sipa** [comment]
I think this is not ready for merge.

**Comment by jonasnick** [comment]
@luke-jr I changed this to draft after we realized that this PR isn't satisfactory, requires more work and has a dependency on another potential BIP-340 update. If you prefer, we can (temporarily) close this PR and move the discussion elsewhere.

**Comment by 0racl3z** [comment]
Still aiming for a solution. I don't recall drafting it

**Comment by gmaxwell** [comment]
An alternative might just be removing the speedup details from the BIP entirely (or just for now).

**Comment by 0racl3z** [comment]
Understood, i'm trying to restore old files in the hopes of any type of
outcome.

**Comment by real-or-random** [comment]
> An alternative might just be removing the speedup details from the BIP entirely (or just for now).

I like that idea. This BIP a spec, which should be valid for years. So at some point the numbers will be outdated anyway... (Yeah, we have also other "motivational" sections which will be outdated at some point but I still don't think we need the speedup numbers.)

**Comment by gmaxwell** [comment]
Should be sufficient for the purpose of the BIP to just say that batch validation is a substantial speedup.

**Review by 0racl3z — APPROVED** [review]

**Review by 0racl3z — APPROVED** [review]
Thx!

**Comment by jonasnick** [comment]
Given that there are apparently many factors that significantly affect the speedup I agree that this plot is not very useful to a reader. We don't even mention that the graph only applies to a specific implementation. And for it's limited usefulness the graph draws too much attention in the BIP.

I'd suggest to add the graph to the libsecp `doc/` dir and link to it from this BIP to prove that batching does in fact lead to a substantial speedup.

**Comment by real-or-random** [comment]
> I'd suggest to add the graph to the libsecp `doc/` dir and link to it from this BIP to prove that batching does in fact lead to a substantial speedup.

That's a good idea.

**Comment by 0racl3z** [comment]
I'm in the process of a solution perhaps. I'll update when an outcome
happens

**Pull request closed without merging**